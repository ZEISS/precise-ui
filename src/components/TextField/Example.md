**Elementary**

Various versions of the text field. By default, we get a plain text field that is managed and enabled. No placeholder is shown. The width of the text field is the width of the parent container.

```jsx
const { TextField } = require('precise-ui');

<TextField />
```

**Decoration**

The `label` prop is used to determine what text should be shown as a label.

```jsx
const { TextField } = require('precise-ui');

<TextField label="Label" />
```

The `placeholder` prop is used to determine what text should be shown as a placeholder (for the empty text field).

```jsx
const { TextField } = require('precise-ui');

<TextField label="With label" placeholder="With placeholder" />
```

The `disabled` prop is indicating that the value cannot be edited, because the control is disabled. There is no `readonly` prop, as such a prop is given indirectly via not updating the `value` prop in controlled mode.

```jsx
const { TextField } = require('precise-ui');

<TextField disabled placeholder="disabled"/>
```

The `clearable` prop is showing a little icon to reset the `TextField` as soon as the user enters data.

```jsx
const { TextField } = require('precise-ui');

<TextField clearable placeholder="Enter something ..."/>
```

**Controlled Mode**

The controlled mode is triggered via the `value` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField value="With value" placeholder="With value"/>
```

You can also add a callback to the clear button by using the `onClear` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField value="With value" placeholder="With value" clearable onClear={() => alert('Clear Pressed')}/>
```

This enables us to also fully utilize a controlled mode. Note, that not only the `onClear` is fired when clearing, but also the `onChange` as we want to change to a new value.

```jsx
const { TextField } = require('precise-ui');

class ControlledClearableTextField extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        value: '',
    };
    this.changed = ({ value }) => {
      this.setState({
          value,
      });
    };
  }

  render() {
      return <TextField clearable value={this.state.value} onChange={this.changed} placeholder="Enter something ..."/>;
}
}

<ControlledClearableTextField />
```

Likewise, we can stay in managed mode by using the `defaultValue` instead.

```jsx
const { TextField } = require('precise-ui');

<TextField defaultValue="With value" placeholder="With default value"/>
```

**Decoration Options**

The text field can be decorated with several options, e.g., adding a `prefix` or `suffix`.

```jsx
const { TextField, Icon } = require('precise-ui');

<div>
  <TextField label="Full Name" prefix={<Icon name="Person" width="100%" height="100%" />} />
  <br />
  <TextField label="E-mail" suffix="@mail.com" />
  <br />
  <TextField label="Website" prefix="https://" suffix=".com" />
</div>
```

We can also indicate errors by placing some text on the `error` prop.

```jsx
const { TextField } = require('precise-ui');

class MyInputForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
    };
    this.changeValue = (e) => {
      this.setState({
        value: e.value,
      });
    };
  }

  render() {
    const { value } = this.state;
    const error = value === 'email' ? '' : 'Value must be "email"';
    return <TextField label="Email" defaultValue="emai" error={error} value={value} onChange={this.changeValue} />;
  }
}

<MyInputForm />
```

**Multiline Input**

Finally, multi-line input is also supported. Once we go into multi-line mode the `prefix` and `suffix` props are no longer supported. By default, the multi-line input is not resizable. This can be changed, however, with the `resizable` prop.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable defaultValue="Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Integer pellentesque quam vel velit. Duis pulvinar. Vivamus porttitor turpisacleo. Curabitur bibendum justo non orci. Integer lacinia."/>
```

You can also add the `label` prop to multiline textfields

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable label="Test label" />
```

You can also add the `label` prop in addition to the placeholder

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable label="Test label" placeholder="Test placeholder" />
```

The resizable can also be restricted to vertical and horizontal.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable="vertical" />
```

The `multiline` prop can be specified as a number, indicating the number of rows initially shown.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline={4} value="My text here" disabled />
```

Last, but not least we can use the `resizable` prop with the `auto` value to automatically adjust the height of the input.

```jsx
const { TextField } = require('precise-ui');

<TextField multiline resizable="auto" defaultValue="My text here" />
```
