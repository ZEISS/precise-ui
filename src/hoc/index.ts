export * from './withResponsive';
export * from './withResponsiveMode';
export * from './withValidator';
export * from './withOutsideClick';
export * from './withFormContext';
