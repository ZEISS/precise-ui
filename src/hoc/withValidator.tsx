import * as React from 'react';

export interface BaseInputProps<TEventArgs> {
  /**
   * Event emitted once the value changes.
   */
  onChange?(e: TEventArgs): void;
  /**
   * The currently displayed error message.
   */
  error?: React.ReactChild;
}

export interface ValidatorProps {
  onSuccess?(): void;
  onError?(): void;
}

export interface ValidatorState {
  error: React.ReactChild | undefined;
}

export function withValidator<TEventArgs, TProps extends BaseInputProps<TEventArgs>>(
  Component: React.ComponentType<TProps>,
  validate: (e: TEventArgs) => React.ReactChild | undefined,
) {
  return class Response extends React.PureComponent<TProps & ValidatorProps, ValidatorState> {
    constructor(props: TProps & ValidatorProps) {
      super(props);
      this.state = {
        error: undefined,
      };
    }

    validate = (e: TEventArgs) => {
      const { onChange, onError, onSuccess } = this.props;
      const error = validate(e);

      if (error !== this.state.error) {
        const notify = error ? onError : onSuccess;

        this.setState({
          error,
        });

        if (typeof notify === 'function') {
          notify();
        }
      }

      if (typeof onChange === 'function') {
        onChange(e);
      }
    };

    render() {
      const { error } = this.state;
      return <Component {...this.props} error={error || this.props.error} onChange={this.validate} />;
    }
  };
}
