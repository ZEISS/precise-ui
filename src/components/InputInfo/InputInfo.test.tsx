import * as React from 'react';
import * as enzyme from 'enzyme';
import { InputInfo } from './';
import { TextField } from '../TextField';
import { withValidator } from '../../hoc/withValidator';

describe('<InputInfo />', () => {
  it('should render <InputInfo> that has an error message', () => {
    const wrapper = enzyme.shallow(<InputInfo>I am wrong</InputInfo>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render an <InputInfo> from jsx', () => {
    const wrapper = enzyme.shallow(<InputInfo />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render an <InputInfo> from direct API without any children', () => {
    const wrapper = enzyme.shallow(React.createElement(InputInfo));
    expect(wrapper).toMatchSnapshot();
  });

  it('should be shown with the withValidator hoc if not changed', () => {
    const RequiredTextField = withValidator(TextField, ({ value }) => !value && 'Input is required');
    const wrapper = enzyme.mount(<RequiredTextField info="Required input" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not be shown with the withValidator hoc if changed', () => {
    const RequiredTextField = withValidator(TextField, ({ value }) => !value && 'Input is required');
    const wrapper = enzyme.mount(<RequiredTextField info="Required input" />);
    (wrapper.instance() as any).validate({
      value: '',
    });
    expect(wrapper).toMatchSnapshot();
  });
});
