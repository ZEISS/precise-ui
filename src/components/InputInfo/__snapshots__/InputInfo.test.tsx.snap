// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputInfo /> should be shown with the withValidation hoc if not changed 1`] = `
.c7 {
  font-size: 0.75rem;
  line-height: 0.875rem;
  color: rgba(114,114,114,1);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  height: 100%;
  position: relative;
  min-width: 0;
  margin: auto;
}

.c5 {
  color: rgba(62,62,62,1);
  background: none;
  border: none;
  border-radius: 0;
  font-size: 1rem;
  font-family: inherit;
  padding: 1rem;
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  width: 100%;
  height: 100%;
  cursor: auto;
  -webkit-appearance: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.c5::-ms-clear {
  display: none;
}

.c5::-webkit-input-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5::-moz-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5:-ms-input-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5::placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5:focus {
  outline: none;
}

.c5:focus::-webkit-input-placeholder {
  opacity: 1;
}

.c5:focus::-moz-placeholder {
  opacity: 1;
}

.c5:focus:-ms-input-placeholder {
  opacity: 1;
}

.c5:focus::placeholder {
  opacity: 1;
}

.c3 {
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  background: rgba(248,248,248,1);
  min-height: 54px;
  max-height: 112px;
  border-bottom: 1px solid rgba(239,239,239,1);
  cursor: auto;
  overflow-y: auto;
}

.c3:hover {
  border-bottom-color: rgba(0,139,208,1);
}

.c6 {
  padding-top: 0.25rem;
  padding-left: 1rem;
  padding-bottom: 0.25rem;
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<Validator
  info="Required input"
>
  <TextField
    info="Required input"
    onChange={[Function]}
  >
    <TextFieldInt
      info="Required input"
      onChange={[Function]}
    >
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-kGXeez",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  position: relative;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-kGXeez",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0"
          >
            <Styled(StackPanel)>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-kpOJdX",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  display: flex;
",
                      ],
                    },
                    "displayName": "Styled(StackPanel)",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-kpOJdX",
                    "target": [Function],
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <StackPanel
                  className="c1"
                >
                  <styled.div
                    className="c1"
                    dir="row"
                    wrapping="nowrap"
                  >
                    <StyledComponent
                      className="c1"
                      dir="row"
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": false,
                            "lastClassName": "c2",
                            "rules": Array [
                              "
  display: flex;
  flex-direction: ",
                              [Function],
                              ";
  flex-wrap: ",
                              [Function],
                              ";
  justify-content: flex-start;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      wrapping="nowrap"
                    >
                      <div
                        className="c1 c2"
                        dir="row"
                      >
                        <styled.div
                          border={1}
                          focused={false}
                          hasValue={false}
                        >
                          <StyledComponent
                            border={1}
                            focused={false}
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-jzJRlG",
                                  "isStatic": false,
                                  "lastClassName": "c3",
                                  "rules": Array [
                                    [Function],
                                  ],
                                },
                                "displayName": "styled.div",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-jzJRlG",
                                "target": "div",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                            hasValue={false}
                          >
                            <div
                              className="c3"
                            >
                              <Component
                                error={false}
                                focused={false}
                                hasValue={false}
                                multiline={false}
                              >
                                <styled.div>
                                  <StyledComponent
                                    forwardedComponent={
                                      Object {
                                        "$$typeof": Symbol(react.forward_ref),
                                        "attrs": Array [],
                                        "componentStyle": ComponentStyle {
                                          "componentId": "sc-gqjmRU",
                                          "isStatic": true,
                                          "lastClassName": "c4",
                                          "rules": Array [
                                            "
  flex-grow: 1;
  display: flex;
  flex-flow: column-reverse;
  height: 100%;
  position: relative;
  min-width: 0;
  margin: auto;
",
                                          ],
                                        },
                                        "displayName": "styled.div",
                                        "foldedComponentIds": Array [],
                                        "render": [Function],
                                        "styledComponentId": "sc-gqjmRU",
                                        "target": "div",
                                        "toString": [Function],
                                        "warnTooManyClasses": [Function],
                                        "withComponent": [Function],
                                      }
                                    }
                                    forwardedRef={null}
                                  >
                                    <div
                                      className="c4"
                                    >
                                      <styled.input
                                        labelShown={false}
                                        onBlur={[Function]}
                                        onChange={[Function]}
                                        onFocus={[Function]}
                                        value=""
                                      >
                                        <StyledComponent
                                          forwardedComponent={
                                            Object {
                                              "$$typeof": Symbol(react.forward_ref),
                                              "attrs": Array [],
                                              "componentStyle": ComponentStyle {
                                                "componentId": "sc-fjdhpX",
                                                "isStatic": false,
                                                "lastClassName": "c5",
                                                "rules": Array [
                                                  "
  ::-ms-clear {
    display: none;
  }

  color: ",
                                                  [Function],
                                                  ";
  background: none;
  border: none;
  border-radius: 0;
  font-size: ",
                                                  "1rem",
                                                  ";
  font-family: inherit;
  padding: ",
                                                  [Function],
                                                  ";
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  width: 100%;
  height: 100%;
  cursor: ",
                                                  [Function],
                                                  ";
  -webkit-appearance: none;
  transition: all 0.2s;

  &::placeholder {
    color: ",
                                                  [Function],
                                                  ";
    opacity: ",
                                                  [Function],
                                                  ";
    transition: inherit;
  }

  &:focus {
    outline: none;

    &::placeholder {
      opacity: 1;
    }
  }
",
                                                ],
                                              },
                                              "displayName": "styled.input",
                                              "foldedComponentIds": Array [],
                                              "render": [Function],
                                              "styledComponentId": "sc-fjdhpX",
                                              "target": "input",
                                              "toString": [Function],
                                              "warnTooManyClasses": [Function],
                                              "withComponent": [Function],
                                            }
                                          }
                                          forwardedRef={[Function]}
                                          labelShown={false}
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          value=""
                                        >
                                          <input
                                            className="c5"
                                            onBlur={[Function]}
                                            onChange={[Function]}
                                            onFocus={[Function]}
                                            value=""
                                          />
                                        </StyledComponent>
                                      </styled.input>
                                    </div>
                                  </StyledComponent>
                                </styled.div>
                              </Component>
                              <Component
                                hasValue={false}
                                onClear={[Function]}
                                onClick={[Function]}
                              />
                            </div>
                          </StyledComponent>
                        </styled.div>
                      </div>
                    </StyledComponent>
                  </styled.div>
                </StackPanel>
              </StyledComponent>
            </Styled(StackPanel)>
            <Component
              bottom="xsmall"
              left="medium"
              top="xsmall"
            >
              <styled.div
                bottom="xsmall"
                left="medium"
                top="xsmall"
              >
                <StyledComponent
                  bottom="xsmall"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kgoBCf",
                        "isStatic": false,
                        "lastClassName": "c6",
                        "rules": Array [
                          [Function],
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kgoBCf",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  left="medium"
                  top="xsmall"
                >
                  <div
                    className="c6"
                  >
                    <Component
                      info="Required input"
                    >
                      <InputInfo>
                        <styled.div>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bdVaJa",
                                  "isStatic": false,
                                  "lastClassName": "c7",
                                  "rules": Array [
                                    "
  ",
                                    "
    ",
                                    "font-size: 0.75rem;
  ",
                                    "
    ",
                                    "line-height: 0.875rem;
  ",
                                    "
    ",
                                    "
  ",
                                    ";
  color: ",
                                    [Function],
                                    ";
",
                                  ],
                                },
                                "displayName": "styled.div",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-bdVaJa",
                                "target": "div",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <div
                              className="c7"
                            >
                              Required input
                            </div>
                          </StyledComponent>
                        </styled.div>
                      </InputInfo>
                    </Component>
                  </div>
                </StyledComponent>
              </styled.div>
            </Component>
          </div>
        </StyledComponent>
      </styled.div>
    </TextFieldInt>
  </TextField>
</Validator>
`;

exports[`<InputInfo /> should not be shown with the withValidation hoc if changed 1`] = `
.c7 {
  font-size: 0.75rem;
  line-height: 0.875rem;
  color: rgba(114,114,114,1);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column-reverse;
  -ms-flex-flow: column-reverse;
  flex-flow: column-reverse;
  height: 100%;
  position: relative;
  min-width: 0;
  margin: auto;
}

.c5 {
  color: rgba(62,62,62,1);
  background: none;
  border: none;
  border-radius: 0;
  font-size: 1rem;
  font-family: inherit;
  padding: 1rem;
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  width: 100%;
  height: 100%;
  cursor: auto;
  -webkit-appearance: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.c5::-ms-clear {
  display: none;
}

.c5::-webkit-input-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5::-moz-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5:-ms-input-placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5::placeholder {
  color: rgba(114,114,114,1);
  opacity: 1;
  -webkit-transition: inherit;
  transition: inherit;
}

.c5:focus {
  outline: none;
}

.c5:focus::-webkit-input-placeholder {
  opacity: 1;
}

.c5:focus::-moz-placeholder {
  opacity: 1;
}

.c5:focus:-ms-input-placeholder {
  opacity: 1;
}

.c5:focus::placeholder {
  opacity: 1;
}

.c3 {
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  background: rgba(248,248,248,1);
  min-height: 54px;
  max-height: 112px;
  border-bottom: 1px solid rgba(239,239,239,1);
  cursor: auto;
  overflow-y: auto;
}

.c3:hover {
  border-bottom-color: rgba(0,139,208,1);
}

.c6 {
  padding-top: 0.25rem;
  padding-left: 1rem;
  padding-bottom: 0.25rem;
}

.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<Validator
  info="Required input"
>
  <TextField
    info="Required input"
    onChange={[Function]}
  >
    <TextFieldInt
      info="Required input"
      onChange={[Function]}
    >
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-kGXeez",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  position: relative;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-kGXeez",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0"
          >
            <Styled(StackPanel)>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-kpOJdX",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  display: flex;
",
                      ],
                    },
                    "displayName": "Styled(StackPanel)",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-kpOJdX",
                    "target": [Function],
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <StackPanel
                  className="c1"
                >
                  <styled.div
                    className="c1"
                    dir="row"
                    wrapping="nowrap"
                  >
                    <StyledComponent
                      className="c1"
                      dir="row"
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": false,
                            "lastClassName": "c2",
                            "rules": Array [
                              "
  display: flex;
  flex-direction: ",
                              [Function],
                              ";
  flex-wrap: ",
                              [Function],
                              ";
  justify-content: flex-start;
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      wrapping="nowrap"
                    >
                      <div
                        className="c1 c2"
                        dir="row"
                      >
                        <styled.div
                          border={1}
                          focused={false}
                          hasValue={false}
                        >
                          <StyledComponent
                            border={1}
                            focused={false}
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-jzJRlG",
                                  "isStatic": false,
                                  "lastClassName": "fvjFQg",
                                  "rules": Array [
                                    [Function],
                                  ],
                                },
                                "displayName": "styled.div",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-jzJRlG",
                                "target": "div",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                            hasValue={false}
                          >
                            <div
                              className="c3"
                            >
                              <Component
                                error={false}
                                focused={false}
                                hasValue={false}
                                multiline={false}
                              >
                                <styled.div>
                                  <StyledComponent
                                    forwardedComponent={
                                      Object {
                                        "$$typeof": Symbol(react.forward_ref),
                                        "attrs": Array [],
                                        "componentStyle": ComponentStyle {
                                          "componentId": "sc-gqjmRU",
                                          "isStatic": true,
                                          "lastClassName": "c4",
                                          "rules": Array [
                                            "
  flex-grow: 1;
  display: flex;
  flex-flow: column-reverse;
  height: 100%;
  position: relative;
  min-width: 0;
  margin: auto;
",
                                          ],
                                        },
                                        "displayName": "styled.div",
                                        "foldedComponentIds": Array [],
                                        "render": [Function],
                                        "styledComponentId": "sc-gqjmRU",
                                        "target": "div",
                                        "toString": [Function],
                                        "warnTooManyClasses": [Function],
                                        "withComponent": [Function],
                                      }
                                    }
                                    forwardedRef={null}
                                  >
                                    <div
                                      className="c4"
                                    >
                                      <styled.input
                                        labelShown={false}
                                        onBlur={[Function]}
                                        onChange={[Function]}
                                        onFocus={[Function]}
                                        value=""
                                      >
                                        <StyledComponent
                                          forwardedComponent={
                                            Object {
                                              "$$typeof": Symbol(react.forward_ref),
                                              "attrs": Array [],
                                              "componentStyle": ComponentStyle {
                                                "componentId": "sc-fjdhpX",
                                                "isStatic": false,
                                                "lastClassName": "c5",
                                                "rules": Array [
                                                  "
  ::-ms-clear {
    display: none;
  }

  color: ",
                                                  [Function],
                                                  ";
  background: none;
  border: none;
  border-radius: 0;
  font-size: ",
                                                  "1rem",
                                                  ";
  font-family: inherit;
  padding: ",
                                                  [Function],
                                                  ";
  box-sizing: border-box;
  box-shadow: none;
  margin: 0;
  width: 100%;
  height: 100%;
  cursor: ",
                                                  [Function],
                                                  ";
  -webkit-appearance: none;
  transition: all 0.2s;

  &::placeholder {
    color: ",
                                                  [Function],
                                                  ";
    opacity: ",
                                                  [Function],
                                                  ";
    transition: inherit;
  }

  &:focus {
    outline: none;

    &::placeholder {
      opacity: 1;
    }
  }
",
                                                ],
                                              },
                                              "displayName": "styled.input",
                                              "foldedComponentIds": Array [],
                                              "render": [Function],
                                              "styledComponentId": "sc-fjdhpX",
                                              "target": "input",
                                              "toString": [Function],
                                              "warnTooManyClasses": [Function],
                                              "withComponent": [Function],
                                            }
                                          }
                                          forwardedRef={[Function]}
                                          labelShown={false}
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          value=""
                                        >
                                          <input
                                            className="c5"
                                            onBlur={[Function]}
                                            onChange={[Function]}
                                            onFocus={[Function]}
                                            value=""
                                          />
                                        </StyledComponent>
                                      </styled.input>
                                    </div>
                                  </StyledComponent>
                                </styled.div>
                              </Component>
                              <Component
                                hasValue={false}
                                onClear={[Function]}
                                onClick={[Function]}
                              />
                            </div>
                          </StyledComponent>
                        </styled.div>
                      </div>
                    </StyledComponent>
                  </styled.div>
                </StackPanel>
              </StyledComponent>
            </Styled(StackPanel)>
            <Component
              bottom="xsmall"
              left="medium"
              top="xsmall"
            >
              <styled.div
                bottom="xsmall"
                left="medium"
                top="xsmall"
              >
                <StyledComponent
                  bottom="xsmall"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kgoBCf",
                        "isStatic": false,
                        "lastClassName": "c6",
                        "rules": Array [
                          [Function],
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kgoBCf",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  left="medium"
                  top="xsmall"
                >
                  <div
                    className="c6"
                  >
                    <Component
                      info="Required input"
                    >
                      <InputInfo>
                        <styled.div>
                          <StyledComponent
                            forwardedComponent={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": Array [],
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-bdVaJa",
                                  "isStatic": false,
                                  "lastClassName": "c7",
                                  "rules": Array [
                                    "
  ",
                                    "
    ",
                                    "font-size: 0.75rem;
  ",
                                    "
    ",
                                    "line-height: 0.875rem;
  ",
                                    "
    ",
                                    "
  ",
                                    ";
  color: ",
                                    [Function],
                                    ";
",
                                  ],
                                },
                                "displayName": "styled.div",
                                "foldedComponentIds": Array [],
                                "render": [Function],
                                "styledComponentId": "sc-bdVaJa",
                                "target": "div",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={null}
                          >
                            <div
                              className="c7"
                            >
                              Required input
                            </div>
                          </StyledComponent>
                        </styled.div>
                      </InputInfo>
                    </Component>
                  </div>
                </StyledComponent>
              </styled.div>
            </Component>
          </div>
        </StyledComponent>
      </styled.div>
    </TextFieldInt>
  </TextField>
</Validator>
`;

exports[`<InputInfo /> should not render an <InputInfo> from direct API without any children 1`] = `""`;

exports[`<InputInfo /> should not render an <InputInfo> from jsx 1`] = `""`;

exports[`<InputInfo /> should render <InputInfo> that has an error message 1`] = `
<styled.div>
  I am wrong
</styled.div>
`;
